### PR TITLE
Fixed edge case with transient persistent read failure for Nexus WF updates and add tests

### PR DESCRIPTION
**Pull request by long-nt-tran**
## What changed?

It seems possible that in a case where:
1. Workflow is closed
2. Completion event is recorded in `UpdateInfos`, but
3. When getting the event from persistence, we get cache miss and persistence read transiently fails

We'd want to propagate the persistence read failure so caller can retry the read later. Without the change in mutable_state_impl.go, we'll instead return a Nexus operation error in `GetNexusUpdateCompletion(...)` with the error string: `"Workflow Update failed because the Workflow completed before the Update completed."`

Added tests for this, and also add more tests for Nexus + WF update scenarios.

## Why?

Hardening feature + add more tests.

## How did you test it?
- [ ] built
- [ ] run locally and tested manually
- [ ] covered by existing tests
- [x] added new unit test(s)
- [x] added new functional test(s)

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -778,6 +781,13 @@ func (ms *MutableStateImpl) GetNexusUpdateCompletion(
...
 	var outcome *updatepb.Outcome
 	if err != nil {
+		// If the completion event ID is recorded but the read failed for a reason other than the
+		// event being absent, the failure is likely transient. Return it so the caller can retry.
+		if !errors.Is(err, errUpdateNotFound) &&
+			!errors.Is(err, errUpdateNotComplete) &&
+			!common.IsNotFoundError(err) {
+			return nexusrpc.CompleteOperationOptions{}, err
+		}
 		// If the workflow is complete but the update outcome is missing we need to respond to all callbacks
```
Comment (long-nt-tran) [inline service/history/workflow/mutable_state_impl.go:791]: once we have nicer fault injection API for persistence fault injection, I can probably write functional test for this instead of just a unit test in mutable_state_impl_test.go like I did in this PR

**Review by mavemuri — APPROVED** [review]
